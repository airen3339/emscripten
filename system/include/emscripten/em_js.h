#ifndef __em_js_h__
#define __em_js_h__

#ifdef __cplusplus
#define _EM_JS_CPP_BEGIN extern "C" {
#define _EM_JS_CPP_END   }
#else // __cplusplus
#define _EM_JS_CPP_BEGIN
#define _EM_JS_CPP_END
#endif // __cplusplus

// EM_JS declares JS functions in C code.

// Implementation details:

// The EM_JS macro is specified as:
//   EM_JS(return type, function name, (arguments), {body})
// but the macro uses __VA_ARGS__ instead of a fourth argument. This is so that
// function bodies containing commas are seamlessly handled.

// EM_JS declares the JS function with a C function prototype, which becomes a
// function import in asm.js/wasm. It also declares an __em_js__-prefixed
// function, which we can use to pass information to the Emscripten compiler
// that survives going through LLVM.
// Example:
//   EM_JS(int, foo, (int x, int y), { return 2 * x + y; })
// would get translated into:
//   int foo(int x, int y);
//   __attribute__((used, visibility("default")))
//   const char* __em_js__foo() {
//     return "(int x, int y)<::>{ return 2 * x + y; }";
//   }
// We pack the arguments and function body into a constant string so it's
// readable from asm.js/wasm post-processing.
// Later we see a function called __em_js__foo, meaning we need to create a JS
// function:
//   function foo(x, y) { return 2 * x + y; }
// We use <::> to separate the arguments from the function body because it isn't
// valid anywhere in a C function declaration.

// Example usages live in tests/core/test_em_js.cpp

#define EM_JS(ret, name, params, ...)          \
  _EM_JS_CPP_BEGIN                             \
  ret name params;                             \
  __attribute__((used, visibility("default"))) \
  const char* __em_js__##name() {              \
    return #params "<::>" #__VA_ARGS__;        \
  }                                            \
  _EM_JS_CPP_END

#endif // __em_js_h__
