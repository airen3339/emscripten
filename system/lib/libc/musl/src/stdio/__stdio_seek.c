#include "stdio_impl.h"

off_t __stdio_seek(FILE *f, off_t off, int whence)
{
	off_t ret;
#ifdef SYS__llseek
	off_t off_hi =
#ifndef __EMSCRIPTEN__
		off >> 32;
#else
		0;
#endif
	if (syscall(SYS__llseek, f->fd, off_hi, off, &ret, whence)<0)
		ret = -1;
#else
	ret = syscall(SYS_lseek, f->fd, off, whence);
#endif
	return ret;
}
