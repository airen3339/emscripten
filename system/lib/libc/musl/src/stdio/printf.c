#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int iprintf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfiprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vfprintf(FILE *restrict f, const char *restrict fmt, va_list ap);

int __small_printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

