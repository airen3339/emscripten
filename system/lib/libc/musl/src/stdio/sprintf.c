#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int siprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsiprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vsprintf(char *restrict s, const char *restrict fmt, va_list ap);

int __small_sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

