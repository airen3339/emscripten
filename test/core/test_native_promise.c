#include <stdio.h>
#include <emscripten/eventloop.h>
#include <emscripten/em_js.h>

void timer_expired(void* user_data) {
  int promise_id = (intptr_t)user_data;
  printf("timer_expired promise_id=%d\n", promise_id);
  emscripten_promise_resolve(promise_id, (void*)100);
}

void async_work(void* user_data, int promise_id) {
  printf("async_work started: user_data=%ld promise_id=%d\n",
         (intptr_t)user_data,
         promise_id);
  emscripten_set_timeout(timer_expired, 1000, (void*)(intptr_t)promise_id);
}

EM_JS_DEPS(deps, "$getPromise,$runtimeKeepalivePush,$runtimeKeepalivePop");

EM_JS(void, exit_after_promises, (int id1, int id2), {
  err(`exit_after_promises: ${id1} ${id2}`);
  runtimeKeepalivePush();
  getPromise(id1).then((value) => {
    getPromise(id2).then((value2) => {
      err(`promises resolved (${value}, ${value2}); exiting`);
      runtimeKeepalivePop();
    });
  });
});

int main() {
  int id1 = emscripten_promise_create(async_work, (void*)42);
  int id2 = emscripten_promise_create(async_work, (void*)99);
  exit_after_promises(id1, id2);
  printf("main done\n");
  return 0;
}
