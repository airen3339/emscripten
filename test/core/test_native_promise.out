async_work started: user_data=42 promise_id=0
async_work started: user_data=99 promise_id=1
exit_after_promises: 0 1
main done
timer_expired promise_id=0
timer_expired promise_id=1
promises resolved (100, 100); exiting
